packages/pwa-kit-runtime/src/utils/logger.js

```diff
@@ -0,0 +1,155 @@
+        const logMessage = {
+            namespace: `${this.packageName}.${namespace}`,
+            message: message.trim(),
+            ...(additionalProperties && {additionalProperties})
+        }
+
+        if (this.format === 'TEXT') {
+            return JSON.stringify(logMessage)
+        }
+
+        return logMessage
+    }
+
+    /**
+     * Prints a log message with the namespace using the console object method set in the message log level.
+     * @param {string} message - The log message.
+     * @param {LogOptions} [options={}] - Optional message details.
+     */
+    printLog(message, options = {}) {
+        const {level, ...messageDetails} = options
+
+        if (!this.shouldLog(level)) {
+            return
+        }
+
+        const formattedMessage = this.formatLogMessage(message, messageDetails)
+
+        switch (level) {
+            case 'error':
+                console.error(formattedMessage)
+                break
+            case 'warn':
+                console.warn(formattedMessage)
+                break
+            case 'info':
+                console.info(formattedMessage)
+                break
+            case 'debug':
+                console.debug(formattedMessage)
+                break
+            default:
+                console.log(formattedMessage)
+                break
+        }
+    }
+
+    /**
+     * Logs a debug message.
+     * @param {string} message - The debug message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    debug(message, details = {}) {
+        this.printLog(message, {level: 'debug', ...details})
+    }
+
+    /**
+     * Logs an info message.
+     * @param {string} message - The info message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    log(message, details = {}) {
+        this.printLog(message, {level: 'info', ...details})
+    }
+
+    /**
+     * Logs an info message.
+     * @param {string} message - The info message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    info(message, details = {}) {
+        this.printLog(message, {level: 'info', ...details})
+    }
+
+    /**
+     * Logs a warning message.
+     * @param {string} message - The warning message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    warn(message, details = {}) {
+        this.printLog(message, {level: 'warn', ...details})
+    }
+
+    /**
+     * Logs an error message.
+     * @param {string} message - The error message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    error(message, details = {}) {
+        this.printLog(message, {level: 'error', ...details})
+    }
+}
+
+/**
+ * Create a logger instances for each package
+ * @param {string} packageName - The name of the package where the logger is used.
+ */
+export const createLogger = (packageName) => new PWAKITLogger(LOG_LEVEL, LOG_FORMAT, packageName)
```

Comment:
What is the up-side to exporting this single function that is kinda like the factory pattern vs exporting the class itself?

Reply:
I exported the class to make testing easier. 
On the user land, it should be enough with just exporting the `createLogger` but testing would be more involved. I can explore that option if we see the effort is worth it.